File: lib/features/saved_billers/data/saved_billers_repository.dart
import 'dart:convert';

import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/storage/key_value_store.dart';
import 'saved_biller.dart';

abstract interface class SavedBillersRepository {
  SavedBillers restore(String userId);
  Future<void> persist(String userId, SavedBillers value);
}

class StoredSavedBillersRepository implements SavedBillersRepository {
  StoredSavedBillersRepository(this._store);

  final KeyValueStore _store;

  static String _keyFor(String userId) => 'savedBillers.$userId';

  @override
  SavedBillers restore(String userId) {
    final raw = _store.read(_keyFor(userId));
    if (raw == null) return SavedBillers.empty();
    final decoded = jsonDecode(raw) as List<dynamic>;
    return SavedBillers(items: [
      for (final entry in decoded.cast<Map<String, dynamic>>())
        SavedBiller(
          billerId: entry['billerId'] as String,
          account: entry['account'] as String,
          nickname: entry['nickname'] as String,
        ),
    ]);
  }

  @override
  Future<void> persist(String userId, SavedBillers value) =>
      _store.write(_keyFor(userId), jsonEncode([
        for (final s in value.items)
          {
            'billerId': s.billerId,
            'account': s.account,
            'nickname': s.nickname,
          },
      ]));
}

final savedBillersRepositoryProvider = Provider<SavedBillersRepository>(
  (ref) => StoredSavedBillersRepository(ref.watch(keyValueStoreProvider)),
);

File: src/features/saved_billers/data/savedBillersRepository.ts
import type { KeyValueStore } from '../../../core/storage/keyValueStore';
import { emptySavedBillers, type SavedBiller, type SavedBillers } from './savedBiller';

export interface SavedBillersRepository {
  restore(userId: string): SavedBillers;
  persist(userId: string, value: SavedBillers): Promise<void>;
}

const keyFor = (userId: string): string => `savedBillers.${userId}`;

export class StoredSavedBillersRepository implements SavedBillersRepository {
  constructor(private readonly store: KeyValueStore) {}

  restore(userId: string): SavedBillers {
    const raw = this.store.read(keyFor(userId));
    if (raw === null) return emptySavedBillers();
    const decoded = JSON.parse(raw) as SavedBiller[];
    return { items: decoded.map((e) => ({ billerId: e.billerId, account: e.account, nickname: e.nickname })) };
  }

  persist(userId: string, value: SavedBillers): Promise<void> {
    return this.store.write(keyFor(userId), JSON.stringify(value.items));
  }
}
